File: lib/manifest-store.js
'use strict';

/**
 * Safe read/modify/write for messages.json — the Node half.
 *
 * This is the mirror of scripts/manifest_store.py. Two processes in two
 * languages write this file: the Python publisher appends messages, and this
 * server marks them listened. Before 2026-08-08 both did the same unsafe thing:
 *
 *     catch (e) { return { messages: [] }; }      // substitute empty
 *     fs.writeFileSync(MESSAGES_FILE, ...)        // then truncate
 *
 * A torn file therefore erased the history on the next ordinary operation —
 * a publish, or a listener simply tapping "listened". 407 records, untracked
 * by git, one interrupted write from gone.
 *
 * The three rules, identical to the Python side:
 *   1. a corrupt manifest is never "empty" — refuse, never invent
 *   2. writes are atomic — temp file, fsync, rename
 *   3. writers take a lock, and it is an O_EXCL sentinel rather than flock so
 *      that Python and Node actually exclude each other
 *
 * LOCK_SUFFIX and LOCK_STALE_MS must match manifest_store.py or the two
 * languages will hold different locks and serialise nothing.
 */

const fs = require('fs');
const path = require('path');

const LOCK_SUFFIX = '.lock';
const LOCK_STALE_MS = 30_000;
const LOCK_TIMEOUT_MS = 10_000;
const LOCK_POLL_MS = 50;

const BACKUP_DIRNAME = 'backups';
const BACKUP_KEEP = 30;

class ManifestCorrupt extends Error {}
class ManifestLocked extends Error {}

const lockPath = (file) => file + LOCK_SUFFIX;

/** Busy-wait for the lock. Synchronous on purpose: the callers are already sync. */
function acquireLock(file, timeoutMs = LOCK_TIMEOUT_MS) {
  const lock = lockPath(file);
  const deadline = Date.now() + timeoutMs;

  for (;;) {
    try {
      const fd = fs.openSync(lock, 'wx');       // O_CREAT|O_EXCL
      fs.writeSync(fd, `${process.pid} ${Date.now() / 1000}\n`);
      fs.closeSync(fd);
      return lock;
    } catch (e) {
      if (e.code !== 'EEXIST') throw e;
      // A crashed writer must not wedge the portal forever.
      try {
        if (Date.now() - fs.statSync(lock).mtimeMs > LOCK_STALE_MS) {
          try { fs.unlinkSync(lock); } catch (_) {}
          continue;
        }
      } catch (_) {
        continue; // released while we looked
      }
      if (Date.now() >= deadline) {
        throw new ManifestLocked(`another writer has held ${path.basename(lock)} for over ${timeoutMs}ms`);
      }
      // Sleep without async: Atomics.wait on a throwaway buffer.
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, LOCK_POLL_MS);
    }
  }
}

function releaseLock(lock) {
  try { fs.unlinkSync(lock); } catch (_) {}
}

/** Parse, or throw. An absent file is legitimately empty; an unreadable one is not. */
function readManifest(file) {
  if (!fs.existsSync(file)) return { messages: [] };
  let raw;
  try {
    raw = fs.readFileSync(file, 'utf8');
  } catch (e) {
    throw new ManifestCorrupt(`${file} could not be read (${e.message})`);
  }
  let data;
  try {
    data = JSON.parse(raw);
  } catch (e) {
    throw new ManifestCorrupt(
      `${file} exists but did not parse (${e.message}). Refusing to write — ` +
      `doing so would replace the history with an empty file.`
    );
  }
  if (!data || !Array.isArray(data.messages)) {
    throw new ManifestCorrupt(`${file} parsed but is not a manifest.`);
  }
  return data;
}

/** Keep a daily copy. Best effort — a backup failure must never fail an operation. */
function rotateBackup(file) {
  try {
    const dir = path.join(path.dirname(file), BACKUP_DIRNAME);
    fs.mkdirSync(dir, { recursive: true });
    const stamp = new Date().toISOString().slice(0, 10).replace(/-/g, '');
    const daily = path.join(dir, `messages.${stamp}.json`);
    if (!fs.existsSync(daily) && fs.existsSync(file)) {
      fs.copyFileSync(file, daily);
      const all = fs.readdirSync(dir).filter((f) => /^messages\..*\.json$/.test(f)).sort();
      for (const old of all.slice(0, Math.max(0, all.length - BACKUP_KEEP))) {
        try { fs.unlinkSync(path.join(dir, old)); } catch (_) {}
      }
    }
  } catch (_) {}
}

/** Replace the manifest atomically. Caller must hold the lock. */
function writeManifest(file, data) {
  if (!data || !Array.isArray(data.messages)) {
    throw new Error('refusing to write a manifest without a messages array');
  }
  rotateBackup(file);

  const tmp = path.join(path.dirname(file), `.${path.basename(file)}.${process.pid}.tmp`);
  try {
    const fd = fs.openSync(tmp, 'w');
    try {
      fs.writeSync(fd, JSON.stringify(data, null, 2) + '\n');
      fs.fsyncSync(fd);            // bytes on disk before the rename
    } finally {
      fs.closeSync(fd);
    }
    fs.renameSync(tmp, file);      // atomic within the directory
  } finally {
    try { fs.unlinkSync(tmp); } catch (_) {}
  }

  try {                            // durably record the rename itself
    const dfd = fs.openSync(path.dirname(file), 'r');
    try { fs.fsyncSync(dfd); } finally { fs.closeSync(dfd); }
  } catch (_) {}
}

/**
 * Read, mutate, write — all under one lock.
 * `mutate(data)` returns true to commit, false to leave the file untouched.
 */
function updateManifest(file, mutate) {
  const lock = acquireLock(file);
  try {
    const data = readManifest(file);
    if (mutate(data) === false) return data;
    writeManifest(file, data);
    return data;
  } finally {
    releaseLock(lock);
  }
}

module.exports = {
  ManifestCorrupt,
  ManifestLocked,
  acquireLock,
  releaseLock,
  readManifest,
  writeManifest,
  updateManifest,
};

File: scripts/manifest_store.py
"""Safe read/modify/write for messages.json.

WHY THIS EXISTS. On 2026-08-08 the manifest held 407 messages going back to
2026-05-02. It is in .gitignore and untracked, so git has never seen a single
one of them, and both writers shared the same two-step defect:

    except Exception:  manifest = {"messages": []}     # substitute empty
    open(MESSAGES_FILE, "w")                           # then truncate

One interrupted write leaves a torn file. The next publish parses it, fails,
silently starts from an empty manifest, and overwrites. 407 records gone,
exit code 0, the word "OK" printed. Reproduced: 101 records became 1. Under
eight concurrent publishers, 101 became 4 — the losers each read the file while
a winner had it truncated.

Three rules follow, and this module exists to make them unavoidable:

  1. A CORRUPT MANIFEST IS NEVER "EMPTY". Refuse loudly. An empty manifest is a
     claim that nothing was ever said, and this file cannot make that claim.
  2. WRITES ARE ATOMIC. Write a sibling temp file, fsync it, then os.replace —
     a rename within a directory either happens or does not. A reader never
     sees a half-written manifest, so rule 1 stops being reachable at all.
  3. WRITERS TAKE A LOCK. server.js writes this file too. The lock is an
     O_EXCL sentinel rather than flock so both languages can hold the same one.

`server.js` has a mirror of this in lib/manifest-store.js. The two must agree
on LOCK_SUFFIX and on the stale timeout, or they will not exclude each other.
"""

from __future__ import annotations

import json
import os
import shutil
import time
from datetime import datetime, timezone
from pathlib import Path

# Mirrored in lib/manifest-store.js — change both or neither.
LOCK_SUFFIX = ".lock"
LOCK_STALE_SECONDS = 30.0
LOCK_TIMEOUT_SECONDS = 10.0
LOCK_POLL_SECONDS = 0.05

BACKUP_DIRNAME = "backups"
BACKUP_KEEP = 30


class ManifestCorrupt(RuntimeError):
    """The manifest exists and could not be parsed.

    Never caught into an empty manifest. The correct response to "I cannot read
    the history" is to stop, not to declare there was none.
    """


class ManifestLocked(RuntimeError):
    """Another writer held the lock longer than we are willing to wait."""


def _lock_path(path: Path) -> Path:
    return path.with_name(path.name + LOCK_SUFFIX)


def acquire_lock(path: Path, timeout: float = LOCK_TIMEOUT_SECONDS) -> Path:
    """Take the writer lock, or raise.

    O_CREAT|O_EXCL is atomic on every filesystem this runs on, and unlike flock
    it is visible to a Node process doing the same thing.
    """
    lock = _lock_path(path)
    deadline = time.monotonic() + timeout

    while True:
        try:
            fd = os.open(lock, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
            os.write(fd, f"{os.getpid()} {time.time()}\n".encode())
            os.close(fd)
            return lock
        except FileExistsError:
            # A crashed writer must not wedge the voice forever.
            try:
                age = time.time() - lock.stat().st_mtime
                if age > LOCK_STALE_SECONDS:
                    lock.unlink(missing_ok=True)
                    continue
            except FileNotFoundError:
                continue  # released while we looked
            if time.monotonic() >= deadline:
                raise ManifestLocked(
                    f"another writer has held {lock.name} for longer than {timeout:.0f}s"
                )
            time.sleep(LOCK_POLL_SECONDS)


def release_lock(lock: Path) -> None:
    try:
        lock.unlink()
    except FileNotFoundError:
        pass


def read_manifest(path: Path) -> dict:
    """Parse the manifest, or raise ManifestCorrupt. Absent file is legitimately empty."""
    if not path.exists():
        return {"messages": []}
    try:
        with open(path, encoding="utf-8") as f:
            data = json.load(f)
    except Exception as exc:
        raise ManifestCorrupt(
            f"{path} exists but did not parse ({exc}). Refusing to continue — "
            f"writing now would replace the history with an empty file. "
            f"A recent copy is in {path.parent / BACKUP_DIRNAME}/."
        ) from exc

    if not isinstance(data, dict) or not isinstance(data.get("messages"), list):
        raise ManifestCorrupt(f"{path} parsed but is not a manifest ({{'messages': [...]}}).")
    return data


def _rotate_backup(path: Path, manifest: dict) -> None:
    """Keep a daily copy. Best effort — a backup failure must never fail a publish."""
    try:
        backups = path.parent / BACKUP_DIRNAME
        backups.mkdir(exist_ok=True)
        stamp = datetime.now(timezone.utc).strftime("%Y%m%d")
        daily = backups / f"messages.{stamp}.json"
        if not daily.exists() and path.exists():
            shutil.copy2(path, daily)
            keep = sorted(backups.glob("messages.*.json"))[:-BACKUP_KEEP]
            for old in keep:
                old.unlink(missing_ok=True)
    except Exception:
        pass


def write_manifest(path: Path, manifest: dict) -> None:
    """Replace the manifest atomically. Caller must hold the lock."""
    if not isinstance(manifest.get("messages"), list):
        raise ValueError("refusing to write a manifest without a messages list")

    _rotate_backup(path, manifest)

    tmp = path.with_name(f".{path.name}.{os.getpid()}.tmp")
    try:
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(manifest, f, indent=2, ensure_ascii=False)
            f.write("\n")
            f.flush()
            os.fsync(f.fileno())  # the bytes are on disk before the rename
        os.replace(tmp, path)     # atomic: readers see old or new, never half
    finally:
        try:
            tmp.unlink()
        except FileNotFoundError:
            pass

    # Durably record the rename itself, so a power loss cannot resurrect the old file.
    try:
        dfd = os.open(path.parent, os.O_RDONLY)
        try:
            os.fsync(dfd)
        finally:
            os.close(dfd)
    except OSError:
        pass


def append_message(path: Path, entry: dict) -> int:
    """Append one message under the lock. Returns the new record count."""
    lock = acquire_lock(path)
    try:
        manifest = read_manifest(path)
        manifest["messages"].append(entry)
        write_manifest(path, manifest)
        return len(manifest["messages"])
    finally:
        release_lock(lock)
